### DISCUSSION TITLE
Make interface methods with a default implementation visible on derived types

### DISCUSSION THREAD

**Discussion by jamesfoster** [dotnet/csharplang#8244] [Language Ideas]
[sharplab.io example](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwG4EMwAIAeBGLAXiwBEBTMAS1TLgDoBhMM9YMvACgEoBuAWABQGbDgBMRUhWq1GzVmVHd+AwcKzoJUMgHdJVGnCWrMWAEYT0jADYB7LZ17HsAYwvW7CoyoGUobMABm6M5kWACyAJ4Akn4UQSEAPAAqAMpkVgEAfIJYWNoAFhShqekBWCDh0bGBwWTJaRmZMFha2tyCAN45uT0AAngAbFglGVhMLGwORJktOsMNAV49fYPqpgDOwGDBwPOlY3Jsio4Cy7ndyyNlDLb23NOzuldLZ1djt54nAL6Cgr0AzFheuJyPpaOVKjF/PE6qDpHBsgIuqczgCgas4QYDhNPhcesizqiAOyPPTwl7LH4o5Z43JozHgm4eY4PVpkgxeL5AA===)

https://github.com/dotnet/roslyn/issues/74138

I had a look for previous discussion on the topic but didn't find anything. Is this something you'd concider supporting?

To me this feels more natural. Methods and properties declared on an interface must be implemented by non-abstract derived types. However, if the interface gives a "default" implementation, the derived type has the option of implementing it or omitting it and therefore accepting the default implementation.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods

**Comment by KennethHoff** [comment]
What would happen if a class implements two interfaces with conflicting members?

**Reply by jamesfoster** [reply]
I would see that as an error case.

**Reply by 333fred** [reply]
What if you don't know it's an error case until runtime because an assembly that was upgraded added a new DIM, an action that is intended to _not_ be a binary breaking change?

**Comment by HaloFour** [comment]
The details are probably buried in the issues, but the design decision was that default implementations work like explicit implementation and that the members do not appear on the implementing type:

https://github.com/dotnet/csharplang/blob/4ed4090c05fe5f0d66c5766bc24749f4b16eda42/meetings/2017/LDM-2017-03-08.md#inheritance-from-interface-to-class

Interface members aren't inherited, and implementing an interface does not require that the implementing type expose public members of the matching name/signature.

**Reply by jamesfoster** [reply]
Thanks for the link to the language design meeting. Very helpful.

While it's focused on the "what if" scenarios, it would be nice to see the simple case just work. Any ambiguity could be seen as an error case.
